### ISSUE TITLE
error: QTC_PLUGIN_NAME is empty

### ISSUE THREAD

**Issue by spachner** [pasccom/QTestLibPlugin#1]
Hi, 

I am using Qt Creator 4.0.82.
I added the QtCreator.local.pri containing the following

```
# Qt Creator source tree:
isEmpty(QTCREATOR_SOURCES):QTCREATOR_SOURCES=/Users/spachner/localPrograms/QTCreator/qt-creator
# Qt Creator build tree:
isEmpty(IDE_BUILD_TREE):IDE_BUILD_TREE=/Users/spachner/localPrograms/QTCreator/qt-creator-build
```

During qmake I get this error.
`
QTC_PLUGIN_NAME is empty. Maybe you meant to create QTestLibPlugin-master_dependencies.pri?`

Any hints?

**Comment by pasccom** [comment]
Hi,
    I guess the folder containing QTestLibPlugin on your computer is called
"QTestLibPlugin-master", right?
    You need to rename the folder in which QTestLibPlugin is
"QTestLibPlugin" (without -master). Indeed qtcreatorplugin.pri uses the
folder name to determine the dependency file name.
    Your *.pri.local file confiuration looks right.
    Let me know if you have any problems and please note that the current
version was tested only against 3.6.0. I hope I have some time to move
to 4.0 soon. Also are you using CMake or qMake? I personnaly use qMake
and never used CMake.
    Best regards,
Pasccom.

**Comment by spachner** [comment]
Hi,

many thanks for your fast answer. You are right, when removing the "-master" the problem is gone. After disabling shadow-build (other path problems) the compilations starts. Unfortunately it failed due to a missing include:

<...>/QTCreatorPlugins/QTestLibPlugin/plaintextqtestlibparserfactory.cpp:24: error: 'projectexplorer/localapplicationrunconfiguration.h' file not found
#include <projectexplorer/localapplicationrunconfiguration.h>
         ^

localapplicationrunconfiguration.h really does not exist in the QTCreator sources. Seems that the new QTCreator is not compatible to your plug-in. Hope you have sometime to dig in. Today the QTCreator 4.0 were officially released. I am really keen using your plug-in.

regards

spachner

**Comment by pasccom** [comment]
Hello,
    I had a look to the new Qt Creator 4.0.0 sources. The run configuration
stuff on which my plugins heavily relies have been fully rewamped (new
classes and style C++0x->C++11).
    I will need some time to fix it. No simple patch can be done. Stay in
touch for next version (0.0.4).
    Regards,
Pasccom.

**Comment by spachner** [comment]
Hi,

I step back to QTCreator 3.6 und got your plugin working. Your delivered tests work, too.

Unfortunately my QT console application based on QTest does not show up in your plugin. The application output looks like this:

Starting /Users/spachner/Documents/svn2/silentbrick/t0/silentbrick/sw/app/unittests/build-QBTest-Qt_5_5_1_clang_64-Debug/QBTest...
********\* Start testing of TestBasics *********
Config: Using QtTest library 5.5.1, Qt 5.5.1 (x86_64-little_endian-lp64 shared (dynamic) release build; by Clang 6.0 (clang-600.0.56) (Apple))
PASS   : TestBasics::initTestCase()
PASS   : TestBasics::testSetPartitionStartLBA()
PASS   : TestBasics::cleanupTestCase()
Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted
********\* Finished testing of TestBasics *********
Info : Test summary: No Errors
/Users/spachner/Documents/svn2/silentbrick/t0/silentbrick/sw/app/unittests/build-QBTest-Qt_5_5_1_clang_64-Debug/QBTest exited with code 0

So far I understand the plugin is simply parsing the QTCreator application window output, there is no special “communication” to the plugin I have to built-in, right? So, what is the problem?

regards

Stefan

**Comment by pasccom** [comment]
Hello,
    I don't think you made any error while building the plugin. If the
provided tests pass, then the build should be right.
    The problem is that the plugin first parses the test project file to
see if the executable being run is really a test. Is your project a
qMake project? Does it defines QT+=testlib? (See the heuristics
described in baseqmakeqtestlibparserfactory.h)
    If my guess is not right, then can you please send me your project
file? Also try getting the last version (v0.0.3).
    If it is right, then one quick fix is to make
BaseQMakeQTestLibParserFactory::canParse(ProjectExplorer::RunConfiguration
*runConfiguration) always return true. The issue with this kind of fix
is that the output of every application you run will be parsed , which
consumes uselessly some ressources. The longer fix is to define an
AbstractTestParserFactory  customized to your special test framework.
This can be a bit tedious if you have never developped plugins for Qt
Creator.
    Having seen your problem and planing there will be quite a lot of
people having a similar one, I will develop an AbstractTestParserFactory
which can be forced by the used through the GUI. Unfortunately, I will
not be able to do this today.
    Regards,
Pasccom.

**Comment by spachner** [comment]
Hi Pasccom,

many thanks for your elaborated answer. I compared my .pro with a working example and do no see any significant difference.

————————————
QT += core
QT -= gui
QT += testlib
QT += serialport

CONFIG += c++11

TARGET  = QBTest
CONFIG += console
CONFIG -= app_bundle

DEFINES += APPLICATION_NAME=\\"QBTest\\"
DEFINES += APPLICATION_VERSION=\\"0.1\\"
DEFINES += DEBUG_WITH_UART_OUT DEBUG_WITH_LEVEL_DEFAULT    # needed for log.h used by C modules

INCLUDEPATH += ../../../modules/

TEMPLATE = app

SOURCES += main.cpp \
    testBasics.cpp \
    testMain.cpp \
    sbserial.cpp \
    sbtransport.cpp \
    sbbasiccmd.cpp \
    sbwormcmd.cpp \
    sbpktcode.cpp \
    ../../../modules/crc16/source/crc16.c \
    ../../../modules/crc16/source/crc16Helper.c \
    testMessageHandler.cpp \
    ../../../modules/commands/source/commands.c \
    ../../../modules/commands/source/commandsBasics.c \
    ../../../modules/commands/source/commandsByHost.c \
    ../../../modules/commands/source/commandsSetRepair.c \
    ../../../modules/commands/source/commandsByEFM.c \
    ../../../modules/aes/host/source/aesImpl.c \
    ../../../modules/aes/host/source/getLTAKeyImpl.c \
    ../../../modules/timer/host/clockImpl.c \
    ../../../modules/aes/common/setRepairSignature.c \
    ../../../modules/timer/common/delay.c \
    ../../../modules/aes/common/aescbc.c

HEADERS += \
    testBasics.h \
    testMain.h \
    sbserial.h \
    sbtransport.h \
    sbbasiccmd.h \
    sbwormcmd.h \
    sbpktcode.h \
    ../../../modules/crc16/include/crc16.h \
    testMessageHandler.h

—————————————— 

I also tried to compile tag 0.0.3 unfortunately without success. I am using QtCreator.local.pri containing:

—————————————— 

# Qt Creator source tree:

isEmpty(QTCREATOR_SOURCES):QTCREATOR_SOURCES=/Users/spachner/localPrograms/qt-creator-3.6.1

# Qt Creator lib path:

isEmpty(IDE_BUILD_TREE):IDE_BUILD_TREE=/Users/spachner/localPrograms/qt-creator-3.6.1

# To compile tests the path to make is needed.

# You can force it here.

FORCE_MAKE_EXE=/usr/bin/make

# To compile tests you may need to provide the path to make

# if it cannot be auto-detected in your system PATH and not

# forced above

MAKE_EXE=/usr/bin/make

—————————————— 

Stefans-iMac-FAST:QTestLibPlugin spachner$ qmake
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]
    [-e pattern] [-f file] [--binary-files=value] [--color=when]
    [--context[=num]] [--directories=action] [--label] [--line-buffered]
    [--null] [pattern] [file ...]
Info: creating stash file /Users/spachner/localPrograms/QTCreatorPlugins/QTestLibPlugin/.qmake.stash
Stefans-iMac-FAST:QTestLibPlugin spachner$ make
cd src/QTestLibPlugin/ && ( test -e Makefile || /Users/spachner/Qt/5.5/clang_64/bin/qmake /Users/spachner/localPrograms/QTCreatorPlugins/QTestLibPlugin/src/QTestLibPlugin/QTestLibPlugin.pro -o Makefile ) && /Applications/Xcode.app/Contents/Developer/usr/bin/make -f Makefile 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -O2 -fPIC -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -std=c++11 -stdlib=libc++ -mmacosx-version-min=10.7 -Wall -W -DQTESTLIBPLUGIN_LIBRARY -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I../../../../qt-creator-3.6.1/src/libs -I../../../../qt-creator-3.6.1/src/plugins -I../../../../qt-creator-3.6.1/src/shared -I/Users/spachner/Qt/5.5/clang_64/lib/QtWidgets.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtGui.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtNetwork.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtCore.framework/Headers -I.moc -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/AGL.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/mkspecs/macx-clang -F/Users/spachner/Qt/5.5/clang_64/lib -o .obj_unix/testoutputpane.o testoutputpane.cpp

… many successful compilations…

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -O2 -fPIC -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -std=c++11 -stdlib=libc++ -mmacosx-version-min=10.7 -Wall -W -DQTESTLIBPLUGIN_LIBRARY -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I../../../../qt-creator-3.6.1/src/libs -I../../../../qt-creator-3.6.1/src/plugins -I../../../../qt-creator-3.6.1/src/shared -I/Users/spachner/Qt/5.5/clang_64/lib/QtWidgets.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtGui.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtNetwork.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtCore.framework/Headers -I.moc -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/AGL.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/mkspecs/macx-clang -F/Users/spachner/Qt/5.5/clang_64/lib -o .obj_unix/plaintextqtestlibparserfactory.o plaintextqtestlibparserfactory.cpp
In file included from plaintextqtestlibparserfactory.cpp:19:
In file included from ./plaintextqtestlibparserfactory.h:22:
In file included from ./baseqmakeqtestlibparserfactory.h:23:
./qtestlibargsparser.h:466:106: error: implicit instantiation of undefined template 'QPair<QString, QStringList>'
    inline void addTestCase(const QString& function, const QStringList& dataTags) {mSelectedTestCases << QPair<QString, QStringList>(function, dataTags);}
                                                                                                         ^
/Users/spachner/Qt/5.5/clang_64/lib/QtCore.framework/Headers/qmetatype.h:2048:1: note: template is declared here
QT_FOR_EACH_AUTOMATIC_TEMPLATE_2ARG(Q_FORWARD_DECLARE_METATYPE_TEMPLATE_2ARG_ITER)
^
/Users/spachner/Qt/5.5/clang_64/lib/QtCore.framework/Headers/qmetatype.h:204:7: note: expanded from macro 'QT_FOR_EACH_AUTOMATIC_TEMPLATE_2ARG'
    F(QPair, struct)
      ^
1 error generated.
make[1]: **\* [.obj_unix/plaintextqtestlibparserfactory.o] Error 1
make: **\* [sub--Users-spachner-localPrograms-QTCreatorPlugins-QTestLibPlugin-src-QTestLibPlugin-make_first-ordered] Error 2

—————————————— 

I added the obvious missing “#include <QPair>” to qtestlibargsparser.h and the compliation went ahead until this:

—————————————— 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -O2 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7 -Wall -W -fPIC -DQT_NO_DEBUG -DQT_TESTLIB_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/Users/spachner/localPrograms/QTCreatorPlugins/QTestLibPlugin/tests/LimitsTest"' -I. -I/Users/spachner/Qt/5.5/clang_64/lib/QtTest.framework/Headers -I/Users/spachner/Qt/5.5/clang_64/lib/QtCore.framework/Headers -I.moc -I/Users/spachner/Qt/5.5/clang_64/mkspecs/macx-clang -F/Users/spachner/Qt/5.5/clang_64/lib -o .obj_unix/limitstest.o limitstest.cpp
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -headerpad_max_install_names -Wl,-syslibroot,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7 -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -weak_framework XCTest -Wl,-rpath,/Users/spachner/Qt/5.5/clang_64/lib -o release/LimitsTest .obj_unix/limitstest.o -F/Users/spachner/Qt/5.5/clang_64/lib -framework QtTest -framework Security -framework ApplicationServices -framework IOKit -framework Foundation -framework QtCore -framework DiskArbitration
cd PlainTextQTestLibParserTest/ && ( test -e Makefile || /Users/spachner/Qt/5.5/clang_64/bin/qmake /Users/spachner/localPrograms/QTCreatorPlugins/QTestLibPlugin/tests/PlainTextQTestLibParserTest/PlainTextQTestLibParserTest.pro -spec macx-clang CONFIG+=x86_64 -o Makefile ) && /Applications/Xcode.app/Contents/Developer/usr/bin/make -f Makefile
make[2]: **\* No rule to make target `../common/debug/libtestcommon.a', needed by`release/plaintextqtestlibparsertest'. Stop.
make[1]: **\* [sub-PlainTextQTestLibParserTest-make_first-ordered] Error 2
make: **\* [sub--Users-spachner-localPrograms-QTCreatorPlugins-QTestLibPlugin-tests-make_first-ordered] Error 2
13:34:27: The process "/usr/bin/make" exited with code 2.
Error while building/deploying project QTestLibPlugin (kit: Qt 5.5.1 (clang_64))
When executing step "Make"
—————————————— 

Now I am stuck.

regards

Stefan

**Comment by spachner** [comment]
Hi Pasccom,

I finally get my project work under QTestLibPlugin 0.0.2 with QTCreator 3.6. The problem seems, that my project uses it own command line parsing. A self-made command line switch passes it arguments to QTest. I removed my command line parsing and pass now everything plain to QTest, and finally the my test output show up in your plugin! I do not know what the problem was caused behind the scenes. I thought the plugin just parses the application output.

I came to that point while trying an other QTest plugin named “AutoTest” which comes with QTCreator 4.0. I feel that it is almost the same as you plugin doing. My original project (with own command line parsing) did not work with that plugin either. When removing my command line stuff the plugin “AutoTest” worked immediately. I was happy to find that and went back to QTCreator 3.6 and found as already said that your plugin is now working too.

Well, many thanks for your help. Since the plugin “AutoTest” works with QTCreator 4.0 will continue work with that.

When you need my help for your work, please let me known.

regards

Stefan

**Comment by pasccom** [comment]
Hi Stefan,

```
I did not yet notice the reapearance of a test parser in Qt Creator. I
```

know I existed once upon a time (7 years ago, see github)
and had been removed abruptly. It was very basic at that time. I'm very
happy to see it come back, even though I thought it would never, because
of some (strange) development considerations.
    For the future, I will correct the few bugs you identified in my
plugin, so that people with Qt Creator <4.0 can still have something
working.
    It seems that the first bug (missing #include <QPair>) is dependant on
Qt framework version I suppose, because I don't reproduce it. The second
one, comes from the priority between debug vs. release build, which I
thought was the opposite...
    Regards,
Pasccom.
